File: Algorithms/Sorting/Count Sort/Count_Sort.java
import java.util.Arrays;

public class Count_Sort {

	public static void main(String[] args) {
		int arr[] = { 9, 6, 3, 5, 3, 4, 3, 9, 6, 4, 6, 5, 8, 9, 9 };
		System.out.println("Array before Sorting: " + Arrays.toString(arr));

		int n = arr.length;
		int min = Integer.MAX_VALUE;
		int max = Integer.MIN_VALUE;
		for (int i = 0; i < n; i++) {
			min = Math.min(min, arr[i]);
			max = Math.max(max, arr[i]);
		}
		int range = max - min + 1;
		System.out.println("Minimum Element = " + min + "\nMaximum Element = " + max + "\nRange = " + range);

		countSort(arr, min, max);
		System.out.println("Array after Sorting: " + Arrays.toString(arr));
	}

	static void countSort(int arr[], int min, int max) {
		int range = max - min + 2;
		int[] farr = new int[range];

		for (int i = 0; i < arr.length; i++) {
			int index = arr[i] - min;
			farr[index]++;
		}

		for (int i = 0; i < farr.length; i++) {
			if (i == 0)
				farr[i] = farr[i] - 1;
			else
				farr[i] = farr[i] + farr[i - 1];
		}

		int[] ans = new int[arr.length];
		for (int i = arr.length - 1; i >= 0; i--) {
			int val = arr[i];
			int index = farr[val - min];
			ans[index] = val;
			farr[val - min]--;
		}

		for (int i = 0; i < arr.length; i++)
			arr[i] = ans[i];
	}
}

File: Algorithms/Sorting/Count Sort/count_sort.cpp
#include <bits/stdc++.h>
using namespace std;

int getMax(int arr[], int size) {
   int max = arr[1];
   for(int i = 2; i<=size; i++) {
      if(arr[i] > max)
         max = arr[i];
   }
   return max; 
}

void countSort(int arr[], int size) {
  int output[size+1];
  int mx=getMax(arr,size);
  int count[mx];

  for (int i = 0; i <= mx; ++i) {
    count[i] = 0;
  }

  for (int i = 0; i < size; i++) {
    count[arr[i]]++;
  }

  for (int i = 1; i <= mx; i++) {
    count[i] += count[i - 1];
  }

  for (int i = size - 1; i >= 0; i--) {
    output[count[arr[i]] - 1] = arr[i];
    count[arr[i]]--;
  }

  for (int i = 0; i < size; i++) {
    arr[i] = output[i];
  }
}

void printArray(int arr[], int size) {
  for (int i = 0; i < size; i++)
    cout << arr[i] << " ";
  cout << endl;
}


int main() {
  int n;
  cin>>n;
  int arr[n];
  for(int i=0;i<n;i++) cin>>arr[i];
  countSort(arr, n);
  printArray(arr, n);
}